### PR TITLE
Fix #5924: don't move static to interfaces

### LINKED ISSUE #5924: Illegal field modifiers in class Matchers: 0x9
The following code produces a load-time exception:

```
Exception in thread "main" java.lang.ClassFormatError: Illegal field modifiers in class Matchers: 0x9
	at java.lang.ClassLoader.defineClass1(Native Method)
```

```Scala
trait Matchers {
  object Helper
}

object Matchers extends Matchers

object Test {
  def main(args: Array[String]): Unit = Matchers
}

```

<details>
<summary>show full log</summary>
<pre>
Exception in thread "main" java.lang.ClassFormatError: Illegal field modifiers in class Matchers: 0x9
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at Test$.main(inner.scala:10)
	at Test.main(inner.scala)
</pre>
</details>

<br/>

This bug is found when working with @cheeseng on ScalaTest macros. Big thanks to @cheeseng for minimizing the problem.

### PR DESCRIPTION
**Pull request by liufengyun**
This is an attempt to fix #5924: don't move static to interfaces

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/MoveStatics.scala [hunk 1]
```diff
@@ -60,6 +61,7 @@ class MoveStatics extends MiniPhase with SymTransformer {
       def move(module: TypeDef, companion: TypeDef): List[Tree] = {
         assert(companion ne module)
         if (!module.symbol.is(Flags.Module)) move(companion, module)
+        else if (companion.ne(null) && companion.symbol.is(Flags.Trait)) List(module, companion)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/MoveStatics.scala]: You can just write `!= null`, it translates to the same code as `ne(null)`. I realize the orginal version uses `ne` as well.

File: compiler/src/dotty/tools/dotc/transform/MoveStatics.scala [hunk 2]
```diff
@@ -24,7 +24,8 @@ class MoveStatics extends MiniPhase with SymTransformer {
   override def phaseName: String = MoveStatics.name
 
   def transformSym(sym: SymDenotation)(implicit ctx: Context): SymDenotation = {
-    if (sym.hasAnnotation(defn.ScalaStaticAnnot) && sym.owner.is(Flags.Module) && sym.owner.companionClass.exists) {
+    if (sym.hasAnnotation(defn.ScalaStaticAnnot) && sym.owner.is(Flags.Module) &&
+        sym.owner.companionClass.exists && !sym.owner.companionClass.is(Flags.Trait)) {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/MoveStatics.scala]: According to https://github.com/lampepfl/dotty/issues/5924#issuecomment-464079014 it's OK to do the move if the field is final so why not add:
```suggestion
        sym.owner.companionClass.exists && !sym.owner.companionClass.is(Flags.Trait, butNot = Final)) {
```

Furthermore, I think that the `OFFSET$` fields generated by the lazy val phase could be marked final in bytecode, so they could still be moved.

File: tests/run/i5924.scala [hunk 3]
```diff
@@ -0,0 +1,9 @@
+trait Matchers {
```
Comment (smarter) [inline tests/run/i5924.scala]: It'd be really good to add some [bytecode tests](https://github.com/lampepfl/dotty/blob/master/compiler/test/dotty/tools/backend/jvm/DottyBytecodeTest.scala) which actually check the behavior of MoveStatics.

**Comment by smarter** [comment]
Also we should add a check in the backend that we never emit a non-final static field in an interface.

File: compiler/src/dotty/tools/dotc/transform/MoveStatics.scala (hunk 2 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/MoveStatics.scala]: Interestingly, `OFFSET$` is actually immutable but changed to mutable in MoveStatics:

https://github.com/lampepfl/dotty/blob/12eef6d8e51dcd8a9a73387659b0f860d7b9ca80/compiler/src/dotty/tools/dotc/transform/MoveStatics.scala#L30

**Comment by liufengyun** [comment]
@smarter Can you give it another review? All feedback are addressed.

File: compiler/test/dotty/tools/backend/jvm/DottyBytecodeTests.scala [hunk 4]
```diff
@@ -695,4 +695,50 @@ class TestBCode extends DottyBytecodeTest {
...
+         |  @static val x = 10
+         |  @static final val y = 10
+         |  @static def f: Int = 30
+         |}
+      """.stripMargin
+
+    checkBCode(source) { dir =>
+      val clsIn   = dir.lookupName("Base.class", directory = false).input
+      val clsNode = loadClassNode(clsIn)
+      getMethod(clsNode, "f")
```
Comment (smarter) [inline compiler/test/dotty/tools/backend/jvm/DottyBytecodeTests.scala]: Would be good to check that the correct flags are set, you can call `.access` on a MethodNode or FieldNode to get its flags: https://static.javadoc.io/org.ow2.asm/asm/6.0/org/objectweb/asm/tree/MethodNode.html#access, and then check if ACC_STATIC and ACC_FINAL are enabled.

**Pull request merged**